### ISSUE TITLE
Calling Product::save can cause unintended price reduction

### ISSUE THREAD

**Issue by seigieu** [thirtybees/thirtybees#1694]
I just found/realized, that `$product->price` property is being overwritten by `Product::getPriceStatic`. 

This can lead to serious problems, like unintended price reduction.
Steps to reproduce
- Must be logged as customer with a discount set on customer group.
- __construct Product with full parameter as true
- __construct will change price property to different than one in database
- Calling save will change this price in databse.
    - Calling save multiple times may reduce each time by discount applied by user.

Few things:
- I think, that Object model is used to access data directly from database and save it. It should not be modified on fly.
- I personally thing, that front end should not edit product object. But there may be cases where it may be needed
- Fixing this one may be tricky, because it will break code that works on this bug. But i think this should be fixed, and developers should switch to `getPrice` method where needed.
- Any mods that will override `Product::getPriceStatic` and change price of product, may lead to saving it to database.
- It's actually impossible to access raw value on front end (if full load parameter is supplied), if user is logged in and in group with discount
- There is `Product::getPrice` method that should be used to fetch product price, if you do not want to access raw value.
- Main problem here is usage of `$use_reduc` as true. But even setting it to false would be wrong (because of mods and maybe other cases).

Relevant line: https://github.com/thirtybees/thirtybees/blob/main/classes/Product.php#L729C36-L729C36

**Comment by getdatakick** [comment]
This is a nasty issue. It will be very hard to fix, and maintain backwards compatibility.

**Comment by seigieu** [comment]
I agree. On one side, removing it may break something. On other side not removing this will make it even more problematic to remove in future. Don't think there is any way to maintain compatibility. Let's look at it that way. We can't keep everything backward compatible.

In this case i think it would be better to cut off this cancer asap. 

Note, that many developers (like I was sure for last 10 years :D ) think its direct database access variable. 
Looking at Prestashop/thitry bees Object model its the only common sense.

Product is most broken of all. It's broken by `$full` parameter, it's broken by `$manufacturer_name` parameter. 
IMHO, $manufacturer_name should be never implemented, but `getManufacturerObject` should be used.

I vote to remove it as the lesser of two evils. Think it's time time end this madness here :smile_cat:

**Comment by getdatakick** [comment]
I would like to remove this as well. But I'm afraid that theme and modules depend extensively upon this functionality. We need to investigate the impact and see how much is this actually used

**Comment by getdatakick** [comment]
My first tests shows good results, looks like core does not use this 'full' functionality much, and themes do not depend on this `price` value.

I suggest we remove this immediately after we release `1.5`, so we have full release to hunt for potential bugs in core and native modules.

**Comment by seigieu** [comment]
@getdatakick You said 1.5 and labeled `Release 1.6`.

That would be good Idea i think.

I checked it does not affect cart. So I would assume it may break things that call on `save` on product from front end  (or even fix bugs that no one ever knew about).

Aside: 
I just tested if this would "fix" (it may be feature, not a bug for some). That if client is in Customer group with 10% off on group. Then PS?TB does not show this rabate as price drop (only prices change with no indication of discount).
However, I can say this does not affect this. So this behaviour is not affected either, if it helps.

Partially shame, cause if somone is discounted by group, it would be good to show them "Hey, you're discounted on evereything cause youre in special group"

**Comment by getdatakick** [comment]
When testing the solution, I discovered one third party module that depends on this functionality. 

Without this, the incorrect price without tax was displayed on front office.

I'm afraid, from BC point of view, we can't remove this. But we should ensure that saving products load in $full=true mode should work properly.

**Comment by gregzawadzki** [comment]
I won't press. However many 3rd party modules use many things that are broken. We can't move forward always trying to keep BC.
Maybe at last add deprecation notice that it will be removed in future. So people will stop using it. 

Bit complex, but remove property, set private $deprecated_price and use _set/_get to display deprecation?

Something like:
```php
<?php

class x {
	
	public function __set($key, $val){
		if($key == 'price') echo "Deprecated write on \$produc->price\r\n";
		
	}
	public function __get($key){
		if($key == 'price') echo "Deprecated read on \$produc->price\r\n";
		
	}
	
}

$x = new x;
echo "Test \r\n";
$x->price = '123';
echo $x->price;

echo "Test \r\n";
$x->price_new = '123';
echo $x->price_new;
```

Output:
```Test 
Deprecated write on $produc->price
Deprecated read on $produc->price
Test 
```

Or it will last forever.
